Code improvements

MicaelJarniac opened the pull request.

MicaelJarniac committed:
Code improvements

MicaelJarniac commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
+latest.json
```
Ignore `latest.json` - that file is automatically generated by the script, and should usually not be shared.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+__version__ = "0.1.0"
+
+
+class Status(str, Enum):
+    ACCEPTED = "Accepted"
+    ACTIVE = "Active"
+    DEFERRED = "Deferred"
+    DRAFT = "Draft"
+    FINAL = "Final"
+    PROVISIONAL = "Provisional"
+    REJECTED = "Rejected"
+    REPLACED = "Replaced"
+    WITHDRAWN = "Withdrawn"
+    SUPERSEDED = "Superseded"
```
Using an `Enum` to store possible PEP statuses. The values match what's returned by their API, so it's possible to do `Status("Accepted")` to get the `Status.ACCESSED` instance, for example. It also inherits from `str`, so they can still be used where a `str` would be expected - mostly.

qexat replied:
perhaps `auto()` and use `Status.<Member>.name.title()` instead?

MicaelJarniac replied:
> perhaps `auto()` and use `Status.<Member>.name.title()` instead?

IMO it's better to do it explicitly like this. Using `name.title()` would obfuscate the connection between the names in the enum and the names in PEP's API.

trag1c replied:
> > perhaps `auto()` and use `Status.<Member>.name.title()` instead?
> 
> IMO it's better to do it explicitly like this. Using `name.title()` would obfuscate the connection between the names in the enum and the names in PEP's API.

How so?

MicaelJarniac replied:
> > > perhaps `auto()` and use `Status.<Member>.name.title()` instead?
> > 
> > 
> > IMO it's better to do it explicitly like this. Using `name.title()` would obfuscate the connection between the names in the enum and the names in PEP's API.
> 
> How so?

Well, for the currently available values, this might work, but this correlation isn't guaranteed by anything I believe. If, for example, one of the values was "In Revision", how would one get to that from `IN_REVISION`? Additional logic would be needed.

IMO it's better to explicitly assign the values to the corresponding names. This might make it easier to read (one can clearly see how the values from the API are being turned into members of the enum) and more future-proof (in case they ever add a new value that doesn't follow the made-up rule of "simply convert from all caps to title case").

trag1c replied:
Using `auto()` is safer because it eliminates the risk of making a typo like `SUPERSEDED = "Superseeded"`. Also DRY :)

MicaelJarniac replied:
I don't really like it much, but I've done it.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+    Status.REJECTED: "&4",
+    Status.REPLACED: "&e",
+    Status.WITHDRAWN: "&d",
+    Status.SUPERSEDED: "",  # FIXME Pick color
```
Added a placeholder for a missing state color.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
 }
-LATEST = Path("pep-tracker/latest.json")
+
+LATEST = Path(__file__).parent.resolve() / "latest.json"
```
Better path handling. This allows the script to be run from outside the project folder, as `python pep-tracker`, and also from inside the project folder, as `python .` or `python __main__.py`. It'll always point to the correct `latest.json` file.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+
+LATEST = Path(__file__).parent.resolve() / "latest.json"
 URL = "https://peps.python.org/api/peps.json"
+PEP = NewType("PEP", str)
```
This is purely for type-hinting and type-checking. The type behaves exactly like a `str`.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+PEP = NewType("PEP", str)
+
+
+def now() -> dt.datetime:
+    return dt.datetime.now(tz=dt.timezone.utc)
```
Extracted this method so it can be reused.

qexat replied:
should be named `now_naive` or `naive_now` imo

MicaelJarniac replied:
> should be named `now_naive` or `naive_now` imo

But it's aware, no?

qexat replied:
> But it's aware, no?

then call it `now_utc` or `utc_now` :trollface:

johnslavik replied:
`utc_now` definitely feels great.

MicaelJarniac replied:
I think it being UTC is irrelevant, as the only important thing about it is that it's aware, and not naive. But I suppose naming it `utc_now` would be fine.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+    return dt.datetime.now(tz=dt.timezone.utc)
+
+
+class StateDict(TypedDict):
+    version: str
+    data: dict[str, str]
+    time: str
```
For type-hinting the `dict` representation of `State`. I'd personally rather use something like attrs + cattrs to do the conversion instead, though.

qexat replied:
why `data`  is not `dict[PEP, Status]`? since `PEP` is essentially a `str`, it's hashable, so I can't think of any issue with doing that way

MicaelJarniac replied:
I think I've tried that, and didn't have a lot of success, but I could try it again.

`Status` won't be automatically recreated, though, and it isn't exactly an `str`. So when I want to go from JSON back to a dict, it'll be read as `str`, and I'll need to manually re-instantiate them. So a `dict[PEP, str]` might work, but a `dict[PEP, Status]` likely won't.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+
     def __init__(
-        self, data: dict[str, str], registered_at: dt.datetime | None = None
+        self, data: dict[PEP, Status], time: dt.datetime | None = None
     ) -> None:
         self._data = data
-        self.date = registered_at or dt.datetime.now(tz=dt.timezone.utc)
+        self.time = time or now()
```
Renamed `date` to `time` (`date` makes it sound like `datetime.date`, and not `datetime.datetime`).

trag1c replied:
> Renamed `date` to `time` (`date` makes it sound like `datetime.date`, and not `datetime.datetime`).

Well `time` makes it sound like `datetime.time`, and not `datetime.datetime`.

MicaelJarniac replied:
What do you think it should be, then?
- `date`
- `time`
- `datetime`
- `dt`
- `when`
- `at`
- Other?

trag1c replied:
`dt` and `datetime` are fine

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
-    def __xor__(self, other: State) -> dict[str, tuple[str, str]]:
-        return {k: (self[k], other[k]) for k in self._data if self[k] != other[k]}
+    def __xor__(self, other: State) -> dict[PEP, tuple[Status, Status]]:
+        # FIXME Handle PEPs that have been newly added, and are only on `other`
```
I believe there's a possible bug where a newly added PEP won't be recognized. I haven't really done anything about it, but at least now there's a comment.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+        # FIXME Handle PEPs that have been newly added, and are only on `other`
+        return {k: (self[k], other[k]) for k in self._data if self[k] is not other[k]}
+
+    @classmethod
+    def from_api(cls, data: dict[str, dict[str, str]], /) -> Self:
+        return cls(data={PEP(k): Status(v["status"]) for k, v in data.items()})
+
+    @classmethod
+    def _migrate(cls, old: dict[str, Any] | StateDict, /) -> StateDict:
+        out = StateDict(version=cls.version, data={}, time=now().isoformat())
+        old_version = old.get("version")
+        if old_version is None:
+            data = cast("dict[str, str]", old.copy())
+            out["data"] = data
+        else:
+            out = cast(StateDict, old)
+        return out
+
+    @classmethod
+    def from_dict(cls, state: StateDict | dict[str, Any], /) -> Self:
+        state_dict = cls._migrate(state)
+        data_raw = state_dict["data"]
+        data = {PEP(k): Status(v) for k, v in data_raw.items()}
+        time = dt.datetime.fromisoformat(state_dict["time"])
+        return cls(data=data, time=time)
+
+    def to_dict(self) -> StateDict:
+        return StateDict(
+            version=self.version,
+            data={str(k): v.value for k, v in self._data.items()},
+            time=self.time.isoformat(),
+        )
```
Converted some module functions to methods of this class. I believe more functions could be turned into methods, but I'm not sure if that'd be desirable.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+    def from_api(cls, data: dict[str, dict[str, str]], /) -> Self:
+        return cls(data={PEP(k): Status(v["status"]) for k, v in data.items()})
+
+    @classmethod
+    def _migrate(cls, old: dict[str, Any] | StateDict, /) -> StateDict:
+        out = StateDict(version=cls.version, data={}, time=now().isoformat())
+        old_version = old.get("version")
+        if old_version is None:
+            data = cast("dict[str, str]", old.copy())
+            out["data"] = data
+        else:
+            out = cast(StateDict, old)
+        return out
```
This method should allow for old versions of `latest.json` to still be recognized, converting them to the newer format.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
         req = requests.get(URL, timeout=15)
-    except (requests.ReadTimeout):
+        req.raise_for_status()
+    except requests.ReadTimeout:
         dexit("&4Connection timed out.")
-    if req.status_code != 200:
+    except HTTPError:
```
I think this is a better way of handling HTTP exceptions - by using `raise_for_status`.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
 
-    def dump(self) -> str:
-        return json.dumps(self._data, indent=2)
+def write_latest(state: State) -> None:
+    LATEST.write_text(json.dumps(state.to_dict(), indent=2))
```
Extracted this bit of duplicate code to a dedicated function.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
     minutes, seconds = divmod(seconds, 60)
 
-    if days > 0:
-        return f"{days} days" if days != 1 else "day"
-    if hours > 0:
-        return f"{hours} hours" if hours != 1 else "hour"
-    if minutes > 0:
-        return f"{minutes} minutes" if minutes != 1 else "minute"
-    return f"{seconds} seconds" if seconds != 1 else "second"
+    if days:
+        return f"{days} days" if days > 1 else "day"
+    if hours:
+        return f"{hours} hours" if hours > 1 else "hour"
+    if minutes:
+        return f"{minutes} minutes" if minutes > 1 else "minute"
+    return f"{seconds} seconds" if seconds > 1 else "second"
```
This function now receives a `timedelta`.

I've broken down the first line into two, which makes it slightly easier to read, and also coincidentally makes it so that this entire first block has the same width, which is visually pleasing.

I've tweaked the logic a bit, but it might not work correctly with negative deltas.

There was a case where `seconds` was between 0 and 1, that it'd print "last 0 seconds", and now it prints "last second".

trag1c replied:
"last 0 seconds" was correct but I like the change 👍

MicaelJarniac commented on __main__.py:
```diff
@@ -90,18 +158,19 @@ def main() -> None:
     if not diff:
         dprint(f"&cNo updates detected in the last {delta}.")
-        sys.exit()
+        return
```
Using `return` instead of `sys.exit()` here not only makes it match the case where `diff` is truthy, but it also results in better output when running the code in interactive mode, with `python -i .`.

With `sys.exit()`, it'd print a traceback when there were no updates.

I've also added another empty `return` to the end of this function because IIRC PEP8 says that you should either have no `return`, or always have `return` for every possibility.

MicaelJarniac commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+from requests.exceptions import HTTPError
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
```
Am importing `typing_extensions.Self` from inside this `if TYPE_CHECKING` block so it doesn't become a runtime dependency.

MicaelJarniac commented:
I've changed the `latest.json` file to now include `version` and `time`.

Including `version` allows it to be migrated to newer formats if needed, and including `time` allows this file to be copied, backed up, uploaded, and downloaded, without breaking the delta time.

To accommodate these changes, the raw data itself is now inside `data`.

I've included a method that automatically migrates from the old to the new format.

trag1c commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
+        # FIXME Handle PEPs that have been newly added, and are only on `other`
+        return {k: (self[k], other[k]) for k in self._data if self[k] is not other[k]}
+
+    @classmethod
+    def from_api(cls, data: dict[str, dict[str, str]], /) -> Self:
+        return cls(data={PEP(k): Status(v["status"]) for k, v in data.items()})
+
+    @classmethod
+    def _migrate(cls, old: dict[str, Any] | StateDict, /) -> StateDict:
+        out = StateDict(version=cls.version, data={}, time=now().isoformat())
+        old_version = old.get("version")
+        if old_version is None:
+            data = cast("dict[str, str]", old.copy())
+            out["data"] = data
+        else:
+            out = cast(StateDict, old)
+        return out
+
+    @classmethod
+    def from_dict(cls, state: StateDict | dict[str, Any], /) -> Self:
```
I don't quite see the reason for making these positional only

MicaelJarniac replied:
I personally tend to prefer positional-only args when a method is super clear about what it wants and has only one arg, but I could easily remove the `/` if you so desire.

trag1c replied:
I would prefer that

trag1c commented on __main__.py:
```diff
@@ -3,81 +3,149 @@
-    except (requests.ReadTimeout):
+        req.raise_for_status()
+    except requests.ReadTimeout:
         dexit("&4Connection timed out.")
-    if req.status_code != 200:
+    except HTTPError:
```
Improve consistency

MicaelJarniac replied:
What do you mean?

trag1c replied:
you explicitly imported `HTTPError` while two lines above have `requests.ReadTimeout`

MicaelJarniac committed:
Don't import HTTPError directly

MicaelJarniac committed:
Unposonlyfied methods

MicaelJarniac committed:
Automatically generate values for `Status` enum

I personally dislike doing it this way, though.

MicaelJarniac committed:
Rename `time` to `datetime`

MicaelJarniac committed:
Suppress type issue caused by bug

https://github.com/python/mypy/issues/7591

https://github.com/python/cpython/issues/86844

johnslavik commented on __main__.py:
```diff
@@ -3,81 +3,154 @@
+
+
+class Status(str, Enum):
+    def _generate_next_value_(  # type: ignore
```
Could you specify which type checker error is being suppressed?

MicaelJarniac replied:
I don't remember exactly, but I think it's related to the one in the comment:
- https://github.com/python/mypy/issues/7591

MicaelJarniac replied:
I suppose I could use a more specific `# type: ignore[???]`.

johnslavik commented on __main__.py:
```diff
@@ -3,81 +3,154 @@
+        name: str, start: int, count: int, last_values: list[str]
+    ) -> str:
+        # https://github.com/python/mypy/issues/7591
+        return name.title().replace("_", " ")
```
Why `.replace("_", " ")`? I don't see a `_` in `Status` members.
The enum itself rather feels unlikely to change.

MicaelJarniac replied:
IIRC I added this to future-proof it if there ever happens to be an item like `FOO_BAR = "Foo Bar"`. But I agree it sounds unlikely, and if it ever happens, this could be re-added.

MicaelJarniac replied:
```suggestion
        return name.title()
```

johnslavik commented on dev-requirements.txt:
```diff
@@ -1,2 +1,3 @@
-isort >= 5.12.0
\ No newline at end of file
+isort >= 5.12.0
+typing_extensions >= 4.5.0
```
```suggestion
```
Completely unnecessary.
See also https://github.com/willmcgugan/declare/pull/3 or https://github.com/python-poetry/poetry/pull/9251.

MicaelJarniac committed:
Update dev-requirements.txt

Co-authored-by: Bartosz Sławecki <bartoszpiotrslawecki@gmail.com>

MicaelJarniac committed:
Update __main__.py
